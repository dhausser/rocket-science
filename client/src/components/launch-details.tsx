import React from "react";
import { gql } from "@apollo/client";

import * as LaunchDetailsTypes from "../pages/__generated__/LaunchDetails";

export const LAUNCH_DATA = gql`
  fragment LaunchDetails on Launch {
    __typename
    flight_number
    mission_name
    mission_id
    launch_year
    launch_date_unix
    launch_date_utc
    launch_date_local
    is_tentative
    tentative_max_precision
    tbd
    launch_window
    rocket {
      rocket_id
      rocket_name
      rocket_type
      first_stage {
        cores {
          core_serial
        }
      }
      second_stage {
        payloads {
          payload_id
        }
      }
    }
    ships
    telemetry {
      flight_club
    }
    launch_site {
      site_id
      site_name
      site_name_long
    }
    launch_success
    links {
      mission_patch
      mission_patch_small
    }
    details
    upcoming
    static_fire_date_utc
    static_fire_date_unix
    timeline {
      webcast_liftoff
    }
    crew {
      members
    }
  }
`;

const LaunchDetails: React.FC<LaunchDetailsTypes.LaunchDetails> = ({
  launch,
}) => {
  return <pre>{JSON.stringify(launch, null, 2)}</pre>;
};

export default LaunchDetails;
