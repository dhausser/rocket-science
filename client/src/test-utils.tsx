import React from 'react';
import { render } from '@testing-library/react';
// this adds custom jest matchers from jest-dom
import '@testing-library/jest-dom/extend-expect';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypename?: any;
  defaultOptions?: any;
  cache?: any;
  resolvers?: any;
  [st: string]: any;
};

const mockLaunch = {
  __typename: 'Launch',
  flight_number: '109',
  mission_name: 'test mission',
  rocket: {
    __typename: 'Rocket',
    rocket_name: 'Rocket name 1',
  },
  launch_site: {
    __typename: 'LaunchSite',
    site_id: 'Launch site id 1',
    site_name: 'Launch site name 1',
    site_name_long: 'Launch site name long 1',
  },
  details: 'test details',
};

const renderApollo = (
  node: any,
  {
    mocks,
    addTypename,
    defaultOptions,
    cache,
    resolvers,
    ...options
  }: RenderApolloOptions = {},
): any => {
  return render(
    <MockedProvider
      mocks={mocks}
      addTypename={addTypename}
      defaultOptions={defaultOptions}
      cache={cache}
      resolvers={resolvers}
    >
      {node}
    </MockedProvider>,
    options,
  );
};

export * from '@testing-library/react';
export { renderApollo, mockLaunch };
