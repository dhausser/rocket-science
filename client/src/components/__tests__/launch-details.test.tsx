import React from 'react';

import { renderApollo, cleanup, waitFor, mockLaunch } from '../../test-utils';
import { GET_LAUNCH_DETAILS, LaunchDetails } from '../launch-details';

const mocks = [
  {
    request: { query: GET_LAUNCH_DETAILS, variables: { id: '109' } },
    result: { data: { launch: mockLaunch } },
  },
];

afterEach(cleanup);

it('renders single launch without error', async () => {
  await renderApollo(<LaunchDetails id="109" />, { mocks });
});

it('should render loading state initially', async () => {
  const { getByText } = await renderApollo(<LaunchDetails id="109" />, {
    mocks,
  });
  waitFor(() => expect(getByText('Loading...')).toBeInTheDocument());
});

it('should render single launch', async () => {
  const { container } = await renderApollo(<LaunchDetails id="109" />, {
    mocks,
  });
  waitFor(() => expect(container).toMatchSnapshot());
});
