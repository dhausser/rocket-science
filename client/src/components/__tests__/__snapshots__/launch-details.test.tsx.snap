// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render single launch 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;

exports[`should render single launch 2`] = `
<div>
  <header>
    test mission
  </header>
  <p>
    Launch site name long 1
  </p>
  <p>
    test details
  </p>
  <pre>
    {
  "__typename": "Launch",
  "flight_number": "109",
  "mission_name": "test mission",
  "launch_site": {
    "__typename": "LaunchSite",
    "site_id": "Launch site id 1",
    "site_name": "Launch site name 1",
    "site_name_long": "Launch site name long 1"
  },
  "details": "test details"
}
  </pre>
  <div
    class="button"
  >
    Back
  </div>
</div>
`;
